liwords-ui/src/App.tsx

```diff
@@ -49,10 +49,18 @@ const App = React.memo(() => {
   useEffect(() => {
     if (redirGame !== '') {
       setRedirGame('');
-      resetStore();
       history.replace(`/game/${encodeURIComponent(redirGame)}`);
+      // reset store later as a separate side effect.
+    }
+  }, [history, redirGame, setRedirGame]);
+
+  const location = useLocation();
+  const knownLocation = useRef(location.pathname); // Remember the location on first render.
+  useEffect(() => {
+    if (knownLocation.current !== location.pathname) {
+      resetStore();
     }
-  }, [history, redirGame, resetStore, setRedirGame]);
+  }, [location.pathname, resetStore]);
```

Comment:
I don't totally understand how useRef works here. Does `knownLocation` ever change after the first render? What if we start at `/` then go to `/game/foo` then go back to `/`? Wouldn't the `if` statement then fail, and the store won't reset?

Reply:
- if we start at `/`, the `knownLocation` for this instance of `<App />` is set to `/`.
- `knownLocation` never changes after the first render.
- the moment we go to `/game/foo`, this instance will `resetStore()`.
- `resetStore()` will unmount this `<App />` and mount a totally new, independent instance of `<App />`.
- when that new instance is first rendered, its `knownLocation` is set to `/game/foo`.
- the old instance will be garbage-collected.
- when we go back to `/`, it is the new instance that will `resetStore()`.

Reply:
got it :+1: